File: server/equipment_power_test.go
package main

import "testing"

func TestEquipmentAcquiredDuringPower(t *testing.T) {
	for _, kind := range []int{3, 4, 5, 6} {
		for gear := 14; gear <= 21; gear++ {
			for _, expireFirst := range []bool{false, true} {
				s := initial()
				p := &s.Players[7]
				attribute := gear - 14
				for i := range p.Stats {
					p.Stats[i] = 173
				}
				collector := 7
				if kind == 3 || kind == 6 {
					collector = 16
				}
				s.pickup(collector, kind)
				powered := kind != 6 || attribute == 3
				powerValue := 250
				if kind == 3 || kind == 6 {
					powerValue = 100
				}
				s.pickup(7, gear)
				if p.Stats[attribute] != 250 {
					t.Fatal("equipment boost missing")
				}
				if expireFirst {
					s.restorePower()
				}
				if p.Stats[attribute] != 250 {
					t.Fatal("expiry removed held equipment")
				}
				if !s.damage(16, 7) {
					t.Fatal("impact did not occur")
				}
				want := 173
				if !expireFirst && powered {
					want = powerValue
				}
				if p.Gear != 0 || p.Stats[attribute] != want {
					t.Fatalf("power %d gear %d expireFirst %v: got %d want %d", kind, gear, expireFirst, p.Stats[attribute], want)
				}
				s.restorePower()
				if p.Stats[attribute] != 173 || p.StatBackup[attribute] != 0 {
					t.Fatal("underlying attribute not restored")
				}
			}
		}
	}
}

func TestPowerAppliedAfterEquipment(t *testing.T) {
	for _, kind := range []int{3, 4, 5, 6} {
		for gear := 14; gear <= 21; gear++ {
			s := initial()
			p := &s.Players[7]
			a := gear - 14
			for i := range p.Stats {
				p.Stats[i] = 173
			}
			s.pickup(7, gear)
			collector := 7
			if kind == 3 || kind == 6 {
				collector = 16
			}
			s.pickup(collector, kind)
			if !s.damage(16, 7) {
				t.Fatal("missing impact")
			}
			want := 173
			if kind != 6 || a == 3 {
				want = 250
				if kind == 3 || kind == 6 {
					want = 100
				}
			}
			if p.Stats[a] != want {
				t.Fatalf("power %d gear %d: got %d want %d", kind, gear, p.Stats[a], want)
			}
			s.restorePower()
			if p.Stats[a] != 173 {
				t.Fatal("underlying value lost")
			}
		}
	}
}

File: src/equipment-power.test.js
import test from "node:test";
import assert from "node:assert/strict";
import {initial} from "./game.js";
import {pickup,damage} from "./features.js";
import {restorePower} from "./attributes.js";

test("equipment acquired during a power restores its underlying attribute across expiry and impact",()=>{
 for(const kind of [3,4,5,6]) for(let gear=14;gear<=21;gear++)
  for(const expireFirst of [false,true]) {
   const s=initial(),p=s.players[7],attribute=gear-14;
   p.stats.fill(173);
   pickup(s,kind===3||kind===6?16:7,kind);
   const powered=kind!==6||attribute===3;
   const powerValue=kind===3||kind===6?100:250;
   pickup(s,7,gear);
   assert.equal(p.stats[attribute],250);
   if(expireFirst) restorePower(s);
   assert.equal(p.stats[attribute],250,"expiry retains held equipment");
   assert.equal(damage(s,16,7),true);
   assert.equal(p.gear,0);
   assert.equal(p.stats[attribute],!expireFirst&&powered?powerValue:173,
    `${kind}/${gear}/${expireFirst}: impact restores the pre-equipment layer`);
   restorePower(s);
   assert.equal(p.stats[attribute],173,"expiry restores the original underlying value");
   assert.equal(p.statBackup[attribute],0);
  }
});

test("power applied after equipment remains active when equipment is knocked off",()=>{
 for(const kind of [3,4,5,6]) for(let gear=14;gear<=21;gear++) {
  const s=initial(),p=s.players[7],a=gear-14;
  p.stats.fill(173);pickup(s,7,gear);
  pickup(s,kind===3||kind===6?16:7,kind);
  const affected=kind!==6||a===3;
  assert.equal(damage(s,16,7),true);
  assert.equal(p.stats[a],affected?(kind===3||kind===6?100:250):173,`${kind}/${gear}`);
  restorePower(s);assert.equal(p.stats[a],173);
 }
});
